src/all/tachidesk/src/eu/kanade/tachiyomi/extension/all/tachidesk/Tachidesk.kt

```diff
@@ -287,186 +326,166 @@ class Tachidesk : ConfigurableSource, UnmeteredSource, HttpSource() {
-            mangaList.addAll(categoryMangaList)
-        }
-
-        // Filter by tags
-        var searchResults = mangaList.toImmutableList()
-        val filterConfigs = mutableListOf<Triple<Boolean, String, List<String>>>()
-        if (tagExcludeList.isNotEmpty()) filterConfigs.add(Triple(false, tagFilterExcludeMode, tagExcludeList))
-        if (tagIncludeList.isNotEmpty()) filterConfigs.add(Triple(true, tagFilterIncludeMode, tagIncludeList))
-        filterConfigs.forEach { config ->
-            val isInclude = config.first
-            val filterMode = config.second
-            val filteredTagList = config.third
-            searchResults = searchResults.filter { mangaData ->
-                val lowerCaseTags = mangaData.genre.map { it.lowercase() }
-                val filterResult = when (filterMode) {
-                    tagModeAndString -> lowerCaseTags.containsAll(filteredTagList.map { tag -> tag.lowercase() })
-                    tagModeOrString -> lowerCaseTags.any { tag -> tag in filteredTagList.map { tag -> tag.lowercase() } }
-                    else -> false
+            // Construct a list of all manga in the required categories by querying each one
+            return apolloClient.value.query(
+                SearchMangaQuery(categoryIdList),
+            )
+                .toFlow()
+                .map { response ->
+                    response.dataAssertNoErrors.mangas.nodes.map { it.mangaFragment }
+                        .distinctBy { it.id }
                 }
-                if (isInclude) filterResult else !filterResult
-            }
-        }
+                .map { mangaList ->
+                    // Filter by tags
+                    var searchResults = mangaList.toList()
+                    val filterConfigs = mutableListOf<Triple<Boolean, String, List<String>>>()
+                    if (tagExcludeList.isNotEmpty()) {
+                        filterConfigs.add(
+                            Triple(
+                                false,
+                                tagFilterExcludeMode,
+                                tagExcludeList,
+                            ),
+                        )
+                    }
+                    if (tagIncludeList.isNotEmpty()) {
+                        filterConfigs.add(
+                            Triple(
+                                true,
+                                tagFilterIncludeMode,
+                                tagIncludeList,
+                            ),
+                        )
+                    }
+                    filterConfigs.forEach { config ->
+                        val isInclude = config.first
+                        val filterMode = config.second
+                        val filteredTagList = config.third
+                        searchResults = searchResults.filter { mangaData ->
+                            val lowerCaseTags = mangaData.genre.map { it.lowercase() }
```

Comment:
you could add the genres as a filter to the gql query

Reply:
Not really possible, since the genre filter is just a string filter, and the actual genre on a manga is a comma delimited string

Reply:
it would look like this
```gql
query MyQuery {
  mangas(
    filter: {genre: {includesInsensitive: "action"}, and/or: {genre: {includesInsensitive: "adventure"}, and/or: { ... }}}
  ) {
    nodes {
      id
    }
  }
}
```

unrelated to this change, but this gets kinda messy, what about updating the StringFilter options (and the other filters as well?) to be list of strings which all have to match? 

which would change the above query to this (must include "action", "adventure" and ...)

```gql
query MyQuery {
  mangas(
    filter: {genre: {includesInsensitive: ["action", "adventure", ...]}}
  ) {
    nodes {
      id
    }
  }
}
```

Reply:
I totally forgot that I can create a MangaFilterInput directly in code, I've found a good way to do this

Reply:
Done